sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/memory.scala

```diff
@@ -143,7 +144,7 @@ case class MemoryStream[A : Encoder](id: Int, sqlContext: SQLContext)
       logDebug(generateDebugString(newBlocks.flatten, startOrdinal, endOrdinal))
 
       newBlocks.map { block =>
-        new MemoryStreamDataReaderFactory(block).asInstanceOf[DataReaderFactory[UnsafeRow]]
+        new MemoryStreamDataReaderFactory(block).asInstanceOf[DataReaderFactory]
```

Comment:
I have seen this pattern many time, the java `List` is a little trouble because it's invariance. Shall we change the interface to use array?

Reply:
cc @rdblue @jose-torres

Reply:
I'd like that, but I don't know if that would make things harder for data source implementers working in Java.

Reply:
Array is a java-friendly type.